torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -123,32 +129,76 @@ def save_state_dict(
         >>>     storage_writer=fs_stroage_writer,
         >>> )
     """
-    (
-        metadata,
-        bytes_write_requests,
-        tensor_write_requests,
-    ) = _prepare(state_dict)
-
-    is_rank0 = not dist.is_initialized() or dist.get_rank(process_group) == 0
-    if is_rank0:
-        storage_writer.prepare()
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
+    data: List[Optional[BaseException]] = [None]
+    if is_coordinator:
+        try:
+            storage_writer.prepare()
+        except BaseException as e:
+            data = [e]
 
     # Writing can only start once prepare has finished
     if dist.is_initialized():
-        dist.barrier(process_group)
-
-    combined_writes: List[Union[TensorWriteRequest, BytesWriteRequest]] = []
-    combined_writes.extend(tensor_write_requests)
-    combined_writes.extend(bytes_write_requests)
+        dist.broadcast_object_list(data, group=process_group, src=coordinator_rank)
+
+    if data[0] is not None:
+        raise CheckpointException("failed to prepare storage", {coordinator_rank : data[0]})
+
+    rank_write_error: Optional[BaseException]
+    try:
+        (
+            metadata,
+            bytes_write_requests,
+            tensor_write_requests,
+        ) = _prepare(state_dict, process_group)
+
+        combined_writes: List[Union[TensorWriteRequest, BytesWriteRequest]] = []
+        combined_writes.extend(tensor_write_requests)
+        combined_writes.extend(bytes_write_requests)
+
+        storage_writer.prepare_storage(combined_writes)
+        bytes_futures = storage_writer.write_bytes(bytes_write_requests)
+        tensor_futures = storage_writer.write_tensors(tensor_write_requests)
+        torch.futures.wait_all([bytes_futures, tensor_futures])
+        rank_write_error = None
+    except BaseException as e:
+        rank_write_error = e
+
+    all_errors: List[Optional[BaseException]]
+    # collect all write errors
+    if dist.is_initialized():
+        all_errors = [None] * dist.get_world_size(process_group)
+        dist.gather_object(
+            obj=rank_write_error,
+            object_gather_list=all_errors if is_coordinator else None,
+            dst=coordinator_rank
+        )
+    else:
+        all_errors = [rank_write_error]
+
+    result: List[Optional[CheckpointException]] = [None]
+    if is_coordinator:
+        message: Optional[str] = None
+        # gather produces an array of arrays, flatten it
+        if any(all_errors):
+            message = "Failed to write data"
+        else:
+            try:
+                storage_writer.write_metadata(metadata=metadata)
+                storage_writer.finish()
```

Comment:
Do we need both write_metadata and finish here?
can this be finish(metadata=metadata)

Reply:
I guess for the narrow protocol we're enforcing right now, it makes sense to merge them.